sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -407,6 +559,411 @@ class GpuDynamicPartitionDataWriter(
+        statsTrackers.foreach(_.closeFile(currentWriterStatus.outputWriter.path()))
+      } finally {
+        currentWriterStatus = null
+      }
+    }
+  }
+
+  /**
+   * convert spillable columnar batch seq to tables and close the input `spills`
+   *
+   * @param spills spillable columnar batch seq
+   * @return table array
+   */
+  def convertSpillBatchesToTables(spills: Seq[SpillableColumnarBatch]): ArrayBuffer[Table] = {
+    withResource(spills) { _ =>
+      val subTablesBuffer = new ArrayBuffer[Table]
+      spills.foreach { spillableCb =>
+        withResource(spillableCb.getColumnarBatch()) { cb =>
+          val currTable = GpuColumnVector.from(cb)
+          subTablesBuffer += currTable
+        }
+      }
+      subTablesBuffer
+    }
+  }
+}
+
+/**
+ * Dynamic partition writer with concurrent writers, meaning multiple concurrent writers are opened
+ * for writing.
+ *
+ * The process has the following steps:
+ *  - Step 1: Maintain a map of output writers per each partition columns. Keep all
+ *    writers opened; Cache the inputted batches by splitting them into sub-groups and
+ *    each partition holds a list of spillable sub-groups; Find and write the max pending
+ *    partition data if the total caches exceed the limitation.
+ *  - Step 2: If number of concurrent writers exceeds limit, fall back to sort-based write
+ *    (`GpuDynamicPartitionDataSingleWriter`), sort rest of batches on partition.
+ *    Write batch by batch, and eagerly close the writer when finishing
+ *    Caller is expected to call `writeWithIterator()` instead of `write()` to write records.
+ *    Note: when fall back to `GpuDynamicPartitionDataSingleWriter`, the single writer should
+ *    restore un-closed writers and should handle un-flushed spillable caches.
+ */
+class GpuDynamicPartitionDataConcurrentWriter(
+    description: GpuWriteJobDescription,
+    taskAttemptContext: TaskAttemptContext,
+    committer: FileCommitProtocol,
+    spec: GpuConcurrentOutputWriterSpec)
+    extends GpuDynamicPartitionDataSingleWriter(description, taskAttemptContext, committer)
+        with Arm with Logging {
+
+  // Keep all the unclosed writers, key is partition directory string.
+  // Note: if fall back to sort-based mode, also use the opened writers in the map.
+  private val concurrentWriters = mutable.HashMap[String, WriterStatusWithCaches]()
+
+  // guarantee to close the caches and writers when task is finished
+  TaskContext.get().addTaskCompletionListener[Unit](_ => closeCachesAndWriters())
+
+  private val outDataTypes = description.dataColumns.map(_.dataType).toArray
+
+  val spillCallback = RapidsBuffer.defaultSpillCallback
+
+  val partitionFlushSize = if (description.concurrentWriterPartitionFlushSize <= 0) {
+    // if the property is equal or less than 0, use default value of parquet or orc
+    val extension = description.outputWriterFactory
+        .getFileExtension(taskAttemptContext).toLowerCase()
+    if (extension.endsWith("parquet")) {
```

Comment:
nit: Could we file a follow on issue to see if there is a cleaner way to do this? The file name ending with a value is not ideal in my opinion, and I thought that we can set configs like this as options to the writer, that this would not pick up.

Reply:
I'll file a follow-on issue after merging this one.

Reply:
Follow-on issue: https://github.com/NVIDIA/spark-rapids/issues/6608